lab11.md

```diff
@@ -0,0 +1,186 @@
+
+```
+$ faas-cli deploy --name env --fprocess="env" --image="functions/alpine:latest" --network=func_functions
+```
+
+* Invoke the function without `--sign` flag:
+
+```
+$ echo "The message" | faas-cli invoke env
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+HOSTNAME=d2c1a2cb20c2
+fprocess=env
+HOME=/root
+Http_X_Call_Id=b84947c6-2970-4fcf-ba3b-66dde6943999
+Http_X_Forwarded_For=10.255.0.2:34974
+Http_X_Forwarded_Host=127.0.0.1:8080
+Http_Content_Length=0
+Http_Accept_Encoding=gzip
+Http_Content_Type=text/plain
+Http_User_Agent=Go-http-client/1.1
+Http_X_Start_Time=1538725657952768349
+...
+```
+
+* Invoke the function again but this time with `--sign` flag:
+
+```
+$ echo -n "The message" | faas-cli invoke env --sign=hmac --key=cookie
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+HOSTNAME=d2c1a2cb20c2
+fprocess=env
+HOME=/root
+Http_User_Agent=Go-http-client/1.1
+Http_Content_Length=0
+Http_Accept_Encoding=gzip
+...
+Http_Hmac=sha1=9239edfe20185eafd7a5513c303b03d207d22f64
+...
+```
+
+We see the `hmac` value transformed to environmental variable `Http_Hmac` with encrypted request `The message` with key `cookie` as `9239edfe20185eafd7a5513c303b03d207d22f64` and encryption method `sha1` appended in front of the value.
+
+## HMAC in action
+
+For our purpose we are going to create a new Python 3 function. Let’s call it `hmac-protected`:
+
+```
+$ faas-cli new --lang python3 hmac-protected --prefix="<your-docker-username>"
+```
+
+Add `payload-secret` which will serve as the key that will encrypt the payload. 
+
+Create `payload-secret` like we did in [lab10](https://github.com/openfaas/workshop/blob/master/lab10.md):
+
+```
+$ echo -n "<your-secret>" | docker secret create payload-secret -
+```
+
+> Note: Remember the string you put in place of  `<your-secret>`
+
+Our `hmac-protected.yml` should look like:
+
+```
+provider:
+  name: faas
+  gateway: http://127.0.0.1:8080
+
+functions:
+  hmac-protected:
+    lang: python3
+    handler: ./hmac-protected
+    image: <your-docker-username>/hmac-protected:latest
+    secrets:
+      - payload-secret
+```
+
+Replace the content of the `handler.py` with the following code:
+
+```
+import os, hmac, hashlib
+
+def validateHMAC(message,secret,hash):
+
+    # The hash starts with the encryption method "sha1=" we need to remove it
+    cleanHash = hash[5:]
+    encodedSecret = secret.encode()
+    encodedMessage = message.encode()
+
+    # Encrypt message with secret
+    expectedMac = hmac.new(encodedSecret,encodedMessage,hashlib.sha1)
+    hmacDigest = expectedMac.hexdigest()
+
+    # Compare the received hash with created hash
+    if hmacDigest == cleanHash:
+        return True
+    return False
+
+def handle(req):
+    # We receive the encrypted message in form of a header
+    encryptedMessage = os.getenv("Http_Hmac")
```

Comment:
Good that you've tried to follow the nomenclature that Alex uses in the go package.

Can we use `messageMAC` rather than `encryptedMessage`. As well as removing 'encrypted' I think this makes the code clearer as we are comparing `messageMAC` with `expectedMAC`